lib/list_versions.go

```diff
@@ -1,119 +1,102 @@
 	if error != nil {
-		return "", error
+		return Release{}, error
 	}
-	// Getting versions from body; should return match /X.X.X/ where X is a number
-	semver := `\/(\d+\.\d+\.\d+)\/?"`
-	r, _ := regexp.Compile(semver)
 	for i := range result {
-		if r.MatchString(result[i]) {
-			str := r.FindString(result[i])
-			trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-			return trimstr, nil
+		if !result[i].IsPrerelease {
+			return result[i], nil
 		}
 	}
-
-	return "", nil
+	return Release{}, nil
 }
 
 //GetTFLatestImplicit :  Get the latest implicit terraform version given the hashicorp url
-func GetTFLatestImplicit(mirrorURL string, preRelease bool, version string) (string, error) {
-
+func GetTFLatestImplicit(mirrorURL string, preRelease bool, version string) (Release, error) {
 	if preRelease == true {
-		//TODO: use GetTFList() instead of GetTFURLBody
-		versions, error := GetTFURLBody(mirrorURL)
+		releases, error := GetTFReleases(mirrorURL, preRelease)
 		if error != nil {
-			return "", error
+			return Release{}, error
 		}
 		// Getting versions from body; should return match /X.X.X-@/ where X is a number,@ is a word character between a-z or A-Z
-		semver := fmt.Sprintf(`\/(%s{1}\.\d+\-[a-zA-z]+\d*)\/?"`, version)
+		semver := fmt.Sprintf(`%s{1}\.\d+\-[a-zA-z]+\d*`, version)
 		r, err := regexp.Compile(semver)
 		if err != nil {
-			return "", err
+			return Release{}, err
 		}
-		for i := range versions {
-			if r.MatchString(versions[i]) {
-				str := r.FindString(versions[i])
-				trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-				return trimstr, nil
+		for i := range releases {
+			if r.MatchString(releases[i].Version) {
+				return releases[i], nil
 			}
 		}
 	} else if preRelease == false {
-		listAll := false
-		tflist, _ := GetTFList(mirrorURL, listAll) //get list of versions
+		tflist, err := GetTFReleases(mirrorURL, preRelease)
 		version = fmt.Sprintf("~> %v", version)
 		semv, err := SemVerParser(&version, tflist)
 		if err != nil {
-			return "", err
+			return Release{}, err
 		}
 		return semv, nil
 	}
-	return "", nil
+	return Release{}, nil
 }
 
-//GetTFURLBody : Get list of terraform versions from hashicorp releases
-func GetTFURLBody(mirrorURL string) ([]string, error) {
+func httpGet(url string, queryParams map[string]string) (*http.Response, error) {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-	hasSlash := strings.HasSuffix(mirrorURL, "/")
-	if !hasSlash { //if does not have slash - append slash
-		mirrorURL = fmt.Sprintf("%s/", mirrorURL)
+	q := req.URL.Query()
+	for k, v := range queryParams {
+		q.Add(k, v)
 	}
-	resp, errURL := http.Get(mirrorURL)
+	req.URL.RawQuery = q.Encode()
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		// should clean this up
+		return nil, errors.New("issue during request")
```

Comment:
Would it make sense to expose error details to the user? Something like this:
```suggestion
		err := fmt.Errorf("issue during request (%d: %q)", res.StatusCode, res.Status)
		return nil, errors.New(err)
```